compiler/luci/pass/src/SubstituteSqueezeToReshapePass.cpp

```diff
@@ -0,0 +1,166 @@
+/*
+ * Copyright (c) 2021 Samsung Electronics Co., Ltd. All Rights Reserved
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "luci/Pass/SubstituteSqueezeToReshapePass.h"
+
+#include <luci/IR/CircleNodes.h>
+
+namespace
+{
+
+/**
+ * @brief Return TRUE if all dim is known && dim > 0 && at least dim 1 exist
+ */
+bool can_squeeze_shape(luci::CircleNode *node)
+{
+  int dim_1_cnt = 0;
+  for (uint32_t r = 0; r < node->rank(); ++r)
+  {
+    if (not node->dim(r).known())
+      return false;
+    if (node->dim(r).value() < 1)
+      return false;
+    if (node->dim(r).value() == 1)
+      dim_1_cnt++;
+  }
+  return dim_1_cnt > 0;
+}
+
+/**
+ * @brief return shape vector of input excluding squeeze_dims
+ */
+std::vector<uint32_t> new_shape_with_squeeze(const luci::CircleNode *input,
+                                             const std::vector<int32_t> &squeeze_dims)
+{
+  std::vector<uint32_t> new_shape;
+
+  if (squeeze_dims.size() == 0)
+  {
+    // prepare new_shape from input shape excluding dim 1
+    for (uint32_t r = 0; r < input->rank(); ++r)
+    {
+      auto v = input->dim(r).value();
+      if (v != 1)
+        new_shape.push_back(v);
+    }
+  }
+  else
+  {
+    // prepare new_shape from input shape excluding squeeze_dims
+    std::vector<uint32_t> temp_shape;
+    for (uint32_t r = 0; r < input->rank(); ++r)
+      temp_shape.push_back(input->dim(r).value());
+    for (auto dim : squeeze_dims)
+    {
+      assert(temp_shape.at(dim) == 1);
+      temp_shape.at(dim) = 0;
+    }
+    for (auto ts : temp_shape)
+      if (ts > 0)
+        new_shape.push_back(ts);
+  }
+
+  // TODO check if new_shape is empty
+
+  return new_shape;
+}
+
+/**
+ * @brief return CircleConst ptr with values of new_shape
+ */
+luci::CircleConst *create_shape_const(loco::Graph *graph, const std::vector<uint32_t> &new_shape)
+{
+  uint32_t dim_size = static_cast<uint32_t>(new_shape.size());
+
+  auto shape_const = graph->nodes()->create<luci::CircleConst>();
+
+  // const shape/dtype
+  shape_const->dtype(loco::DataType::S32);
+  shape_const->shape({dim_size});
```

Comment:
If `new_shape` is empty, it means scalar.
However, scalar( [] ) and [0] is different shape.

```suggestion
  shape_const->rank(dim_size);
  if (dim_size > 0)
    shape_const->shape({dim_size});
```

Reply:
calling `shape_const->shape({dim_size});` will update rank too. I'll revise this.

Reply:
Oh, I wrote wrong suggestion code :(

My point was that if `dim_size` is 0, shape would be [0], not [] :)
Sorry for confusion.

Reply:
we are setting to rank(1) not rank(dim_size) dim(0) value is to be set to dim_size

Reply:
I wanted to describe following example.

```python
# Current code
x = tf.constant(1, shape=[1, 1])
y = tf.squeeze(x)
print(y) # tf.Tensor(1, shape=(), dtype=int32)

# AS-IS (Error)
asis_optimized_y = tf.reshape(x, [0])

# TO-BE
tobe_optimized_y = tf.reshape(x, [])
```

With above example, `new_shape` parameter in `luci::CircleConst *create_shape_const(loco::Graph *graph, const std::vector<uint32_t> &new_shape)` will be empty.
As a result, `dim_size` will be 0.
If we use `shape_const->shape({dim_size})`, the result would be `asis_optimized_y`.

Reply:
I think your example needs to be added to the test too